### PR TITLE
Add error returns to void functions that can fail/set errors.

### PR DESCRIPTION
**Pull request by LiquidityC**
This takes care of the last list of void functions that could potentially be shifted to instead return an int indicating success and setting an error in case of an error.

In line with #7233 this patch updates the last void functions that could instead return an int indicating success or failure as well as setting an error.

## Description
Based on the following list of remaining void functions the following have been either handled or skipped.
This list  is taken from a comment in #7233 

In some cases error handling has been introduced (param validation in all the cases if I'm not mistaken) instead of silently doing nothing. Perhaps this is more then what was needed?

- [ ] _SDL_atomic.h:extern DECLSPEC void SDLCALL SDL_AtomicLock(SDL_SpinLock *lock);_
- [ ] _SDL_atomic.h:extern DECLSPEC void SDLCALL SDL_AtomicUnlock(SDL_SpinLock *lock);_
- [ ] _SDL_audio.h:extern DECLSPEC void SDLCALL SDL_ClearAudioStream(SDL_AudioStream *stream);_
- [ ] _SDL_audio.h:extern DECLSPEC void SDLCALL SDL_DestroyAudioStream(SDL_AudioStream *stream);_
- [x] __SDL_gamepad.h:extern DECLSPEC void SDLCALL SDL_SetGamepadPlayerIndex(SDL_Gamepad *gamepad, int player_index);__
- [ ] _SDL_gamepad.h:extern DECLSPEC void SDLCALL SDL_CloseGamepad(SDL_Gamepad *gamepad);_
- [x] __SDL_guid.h:extern DECLSPEC void SDLCALL SDL_GUIDToString(SDL_GUID guid, char *pszGUID, int cbGUID);__
- [ ] _SDL_haptic.h:extern DECLSPEC void SDLCALL SDL_HapticClose(SDL_Haptic * haptic);_
- [ ] _SDL_haptic.h:extern DECLSPEC void SDLCALL SDL_HapticDestroyEffect(SDL_Haptic * haptic,_
- [ ] _SDL_hidapi.h:extern DECLSPEC void SDLCALL SDL_hid_free_enumeration(SDL_hid_device_info *devs);_
- [ ] _SDL_hints.h:extern DECLSPEC void SDLCALL SDL_DelHintCallback(const char *name,_
- [x] __SDL_joystick.h:extern DECLSPEC void SDLCALL SDL_GetJoystickGUIDString(SDL_JoystickGUID guid, char *pszGUID, int cbGUID);__
- [ ] _SDL_joystick.h:extern DECLSPEC void SDLCALL SDL_GetJoystickGUIDInfo(SDL_JoystickGUID guid, Uint16 *vendor, Uint16 *product, Uint16 *version, Uint16 *crc16);_
- [ ] _SDL_loadso.h:extern DECLSPEC void SDLCALL SDL_UnloadObject(void *handle);_
- [ ] _SDL_log.h:extern DECLSPEC void SDLCALL SDL_Log(SDL_PRINTF_FORMAT_STRING const char *fmt, ...) SDL_PRINTF_VARARG_FUNC(1);_
- [ ] _SDL_log.h:extern DECLSPEC void SDLCALL SDL_LogVerbose(int category, SDL_PRINTF_FORMAT_STRING const char *fmt, ...) SDL_PRINTF_VARARG_FUNC(2);_
- [ ] _SDL_log.h:extern DECLSPEC void SDLCALL SDL_LogDebug(int category, SDL_PRINTF_FORMAT_STRING const char *fmt, ...) SDL_PRINTF_VARARG_FUNC(2);_
- [ ] _SDL_log.h:extern DECLSPEC void SDLCALL SDL_LogInfo(int category, SDL_PRINTF_FORMAT_STRING const char *fmt, ...) SDL_PRINTF_VARARG_FUNC(2);_
- [ ] _SDL_log.h:extern DECLSPEC void SDLCALL SDL_LogWarn(int category, SDL_PRINTF_FORMAT_STRING const char *fmt, ...) SDL_PRINTF_VARARG_FUNC(2);_
- [ ] _SDL_log.h:extern DECLSPEC void SDLCALL SDL_LogError(int category, SDL_PRINTF_FORMAT_STRING const char *fmt, ...) SDL_PRINTF_VARARG_FUNC(2);_
- [ ] _SDL_log.h:extern DECLSPEC void SDLCALL SDL_LogCritical(int category, SDL_PRINTF_FORMAT_STRING const char *fmt, ...) SDL_PRINTF_VARARG_FUNC(2);_
- [ ] _SDL_log.h:extern DECLSPEC void SDLCALL SDL_LogGetOutputFunction(SDL_LogOutputFunction *callback, void __userdata);_
- [ ] _SDL_log.h:extern DECLSPEC void SDLCALL SDL_LogSetOutputFunction(SDL_LogOutputFunction callback, void *userdata);_
- [ ] _SDL_mouse.h:extern DECLSPEC void SDLCALL SDL_WarpMouseInWindow(SDL_Window * window,_
- [ ] _SDL_mouse.h:extern DECLSPEC void SDLCALL SDL_DestroyCursor(SDL_Cursor * cursor);_
- [ ] _SDL_mutex.h:extern DECLSPEC void SDLCALL SDL_DestroyMutex(SDL_mutex * mutex);_
- [ ] _SDL_mutex.h:extern DECLSPEC void SDLCALL SDL_DestroySemaphore(SDL_sem *sem);_
- [ ] _SDL_mutex.h:extern DECLSPEC void SDLCALL SDL_DestroyCond(SDL_cond *cond);_
- [ ] _SDL_rwops.h:extern DECLSPEC void SDLCALL SDL_DestroyRW(SDL_RWops * area);_
- [ ] _SDL_sensor.h:extern DECLSPEC void SDLCALL SDL_CloseSensor(SDL_Sensor *sensor);_
- [ ] _SDL_surface.h:extern DECLSPEC void SDLCALL SDL_DestroySurface(SDL_Surface *surface);_
- [ ] _SDL_surface.h:extern DECLSPEC void SDLCALL SDL_UnlockSurface(SDL_Surface *surface);_
- [x] __SDL_surface.h:extern DECLSPEC void SDLCALL SDL_GetSurfaceClipRect(SDL_Surface *surface,__
- [ ] _SDL_system.h:extern DECLSPEC void SDLCALL SDL_SetWindowsMessageHook(SDL_WindowsMessageHook callback, void *userdata);_
- [ ] _SDL_thread.h:extern DECLSPEC void SDLCALL SDL_WaitThread(SDL_Thread * thread, int *status);_
- [ ] _SDL_thread.h:extern DECLSPEC void SDLCALL SDL_DetachThread(SDL_Thread * thread);_
- [x] __SDL_version.h:extern DECLSPEC void SDLCALL SDL_GetVersion(SDL_version * ver);__

## Existing Issue(s)

#7233

### CODE REVIEW

File: src/joystick/SDL_gamepad.c [hunk 1]
```diff
@@ -2955,48 +2956,49 @@ int SDL_SendGamepadEffect(SDL_Gamepad *gamepad, const void *data, int size)
...
+        if (--gamepad->ref_count > 0) {
+            SDL_UnlockJoysticks();
+            return 0;
+        }
 
-    SDL_CloseJoystick(gamepad->joystick);
+        if (SDL_CloseJoystick(gamepad->joystick) != 0) {
+            /* SDL_SetError() called within SDL_CloseJoystick() */
+            return -1;
+        }
```
Comment (LiquidityC) [inline src/joystick/SDL_gamepad.c:2976]: I'm unsure about the above __if__ block. Do all gamepads have joysticks. My quick skim of the code seemed to point in that direction but I'm not 100%. Some assurance would be nice.

File: include/SDL3/SDL_gamepad.h [hunk 2]
```diff
@@ -1032,12 +1034,14 @@ extern DECLSPEC int SDLCALL SDL_SendGamepadEffect(SDL_Gamepad *gamepad, const vo
...
  *                SDL_OpenGamepad()
+ * \returns 0 on success or a negative error code on failure; call
+ *          SDL_GetError() for more information.
  *
  * \since This function is available since SDL 3.0.0.
  *
  * \sa SDL_OpenGamepad
  */
-extern DECLSPEC void SDLCALL SDL_CloseGamepad(SDL_Gamepad *gamepad);
+extern DECLSPEC int SDLCALL SDL_CloseGamepad(SDL_Gamepad *gamepad);
```
Comment (slouken) [inline include/SDL3/SDL_gamepad.h:1044]: The close/destroy functions should probably still return void. What would the application do if it fails?

Comment (LiquidityC) [reply include/SDL3/SDL_gamepad.h:1044]: Try the same thing again and hope it works this time?

I agree, this one must have slipped through. I ignored the other destruct functions for this particular reason.

File: src/joystick/SDL_gamepad.c (hunk 1 above)
Comment (slouken) [reply src/joystick/SDL_gamepad.c:2976]: They all do, yes, but a failure here shouldn't prevent the rest of the cleanup. See my comment above about close/destroy functions in general.

**Comment by 1bsyl** [comment]
I can do the SDL_audio.h AudioID function

File: include/SDL3/SDL_gamepad.h (hunk 2 above)
Comment (LiquidityC) [reply include/SDL3/SDL_gamepad.h]: Ill reset the haptic and cursor destructors as well. Gimme a few :wink:

Comment (LiquidityC) [reply include/SDL3/SDL_gamepad.h]: Done

**Comment by slouken** [comment]
Can you squash for review?

Thanks!

**Comment by 1bsyl** [comment]
I will add more : 
SDL_AtomicLock, SDL_AtomicUnlock

**Comment by 1bsyl** [comment]
also:
SDL_HapticClose
SDL_HapticDestroyEffect

**Comment by 1bsyl** [comment]
Also SDL_hid_free_enumeration and
SDL_DelHintCallback

**Comment by LiquidityC** [comment]
> also: SDL_HapticClose SDL_HapticDestroyEffect

I had those implemented but reverted them. Generally destroy/close functions are assumed to nicely handle NULL arguments. Secondly, there's not really anything the developer can do if a destroy doesn't work.

**Comment by LiquidityC** [comment]
The reasoning for Lock/Unlock is the same.

**Comment by 1bsyl** [comment]
for:  SDL_AtomicLock, SDL_AtomicUnlock
it seems it can checks for null pointer before. 

SDL_HapticClose SDL_HapticDestroyEffect: 
it seems it does some check about valid haptics ..

SDL_hid_free_enumeration and SDL_DelHintCallback:  no need, indeed

**Comment by 1bsyl** [comment]
ok, so I stop. and will just submit my branch, in case it's worth adding it

**Comment by 1bsyl** [comment]
yes, true, so maybe we should reconsider those ones:

```
SDL_pixels.h:extern int SDLCALL SDL_DestroyPixelFormat(SDL_PixelFormat *format);
SDL_pixels.h:extern DECLSPEC int SDLCALL SDL_DestroyPalette(SDL_Palette * palette);
SDL_render.h:extern DECLSPEC int SDLCALL SDL_DestroyTexture(SDL_Texture *texture);
SDL_render.h:extern DECLSPEC int SDLCALL SDL_DestroyRenderer(SDL_Renderer *renderer);
SDL_video.h:extern DECLSPEC int SDLCALL SDL_DestroyWindow(SDL_Window *window);

SDL_audio.h:extern DECLSPEC int SDLCALL SDL_UnlockAudioDevice(SDL_AudioDeviceID dev);
SDL_mutex.h:extern DECLSPEC int SDLCALL SDL_UnlockMutex(SDL_mutex * mutex) SDL_RELEASE(mutex);
SDL_render.h:extern DECLSPEC int SDLCALL SDL_UnlockTexture(SDL_Texture *texture);

```

**Comment by slouken** [comment]
Yes, let's revert those.

**Comment by 1bsyl** [comment]
Also
`SDL_audio.h:extern DECLSPEC int SDLCALL SDL_CloseAudioDevice(SDL_AudioDeviceID dev);`

I can create a PR for thoses

**Comment by 1bsyl** [comment]
ok. some removed: 
https://github.com/libsdl-org/SDL/pull/7277

except: 
SDL_UnlockMutex because it set relevant erros I guess. (mutex that don t belong to a thread)
SDL_UnlockTexture: cannot unlock an non streaming texture

**Comment by slouken** [comment]
Let's go ahead and revert SDL_UnlockTexture as well. You can't lock a streaming texture, so unlocking it is a no-op (and probably shouldn't set an error)

**Comment by 1bsyl** [comment]
done https://github.com/libsdl-org/SDL/commit/dc74a67a57e4b5ef767201f7ff166af7bf871f21

**Pull request merged**